feat(forms): add `ng-submitted` class to forms that have been submitted.

dylhunn opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: 30486

Currently, no status class is set on a form to show that it has been submitted. As previously discussed in pull/31070 and issues/30486, this would be useful because it is often desirable to apply styles to fields that are both `ng-invalid` and `ng-pristine` after the first attempt at form submission, but Angular does not currently provide any simple way to do this (although evidently Angularjs did).

## What is the new behavior?

A new `ng-submitted` class is set on submitted forms. It will now be possible to select untouched but invalid fields in a submitted form (for example, with a selector such as`.ng-submitted .ng-invalid`).

In this implementation, the directive that sets control status classes on forms and formGroups has its set of statuses widened to include `ng-submitted`. Then, in the event that `is('submitted')` is invoked, the `submitted` property of the control container is returned iff it exists. This is preferred over checking whether the container is a `Form` or `FormGroup` directly to avoid reflecting on those classes.


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information


Closes #30486.
Closes #31070.

AndrewKushnir commented on packages/forms/test/reactive_integration_spec.ts:
```diff
@@ -1123,6 +1129,60 @@ const ValueAccessorB = createControlValueAccessor('[cva-b]');
+        const [loginInput, passwordInput] = fixture.debugElement.queryAll(By.css('input'));
+        const formEl = fixture.debugElement.query(By.css('form')).nativeElement;
+        loginInput.nativeElement.value = 'Nancy';
+        dispatchEvent(loginInput.nativeElement, 'input');
+        dispatchEvent(loginInput.nativeElement, 'blur');
```
Could you please add a quick comment above these lines why we need to invoke both events?

AndrewKushnir commented on packages/forms/test/template_integration_spec.ts:
```diff
@@ -2263,6 +2305,26 @@ class NgModelNgIfForm {
+  `
+})
+class NgModelNestedForm {
+  // TODO(issue/24571): remove '!'.
+  first!: string;
+  // TODO(issue/24571): remove '!'.
+  other!: string;
```
We should remove these TODOs and either keep the `!` (probably the best in our case) or initialize these properties with an empty string.

dylhunn replied:
Sure, I just cargo-culted these from elsewhere in the file.

AndrewKushnir reviewed (commented):
@dylhunn great start, the changes are in the right direction 👍 

There are few comments that I've left + I think we'd need to add more tests (to verify the fact that only the *top-level* form groups receive `ng-submitted` class). I will write a separate comment related to tests and just want to share the rest of the feedback, so you can have a look at it meanwhile.

Thank you.

AndrewKushnir commented:
@dylhunn just wanted to share some test cases that I think would be great to have. Some of them after already fully covered (for ex. "Template-driven Forms" -> Case 2), I listed the cases anyways just to have a full picture (for myself, so I'm not missing anything). I also mentioned what we can look for in the tests, to make sure we do **not** assign extra classes to elements that are not expected to have that class.

## Reactive Forms

### 1. Simple case:

```
<form [formGroup]="group">
 <input [formControl]="control">
</form>
```
We can check that:
- `<form>` has `ng-submitted` class after submission and has no class after reset
- `<input>` has *no* `ng-submitted` class before submission and after reset

### 2. With nested `formGroup`s:

```
<form [formGroup]="group">
   <div [formGroup]="nestedGroup">
     <input [formControl]="control">
   </div>
</form>
```
We can check that:
- `<form>` has `ng-submitted` class after submission and has no class after reset
- `<div>` has *no* `ng-submitted` class before submission and after reset
- `<input>` has *no* `ng-submitted` class before submission and after reset

### 3. With nested `formArrayName`s:

```
<form [formGroup]="group">
   <div [formArrayName]="name">
     <input [formControl]="control">
   </div>
</form>
```
We can check that:
- `<form>` has `ng-submitted` class after submission and has no class after reset
- `<div>` has *no* `ng-submitted` class before submission and after reset
- `<input>` has *no* `ng-submitted` class before submission and after reset

### 4. Similar to a test case above, but with nested `formGroupName`s:

```
<form [formGroup]="group">
   <div [formGroupName]="name">
     <input [formControl]="control">
   </div>
</form>
```
We can check that:
- `<form>` has `ng-submitted` class after submission and has no class after reset
- `<div>` has *no* `ng-submitted` class before submission and after reset
- `<input>` has *no* `ng-submitted` class before submission and after reset


## Template-driven Forms

### 1. Simple case:

```
<form>
 <input [(ngModel)]="control">
</form>
```

We can check that:
- `<form>` has `ng-submitted` class after submission and has no class after reset
- `<input>` has *no* `ng-submitted` class before submission and after reset

### 2. With nested `ngModelGroup`s:

```
<form>
  <div ngModelGroup="group">
    <input [(ngModel)]="control">
  </div>
</form>
```

We can check that:
- `<form>` has `ng-submitted` class after submission and has no class after reset
- `<div>` has *no* `ng-submitted` class before submission and after reset
- `<input>` has *no* `ng-submitted` class before submission and after reset

AndrewKushnir commented:
@dylhunn quick note: we should also update the relevant section of the `aio/content/guide/forms.md` file to include information on the `ng-submitted` class.

AndrewKushnir commented on packages/forms/test/reactive_integration_spec.ts:
```diff
@@ -1123,6 +1138,126 @@ const ValueAccessorB = createControlValueAccessor('[cva-b]');
+        expect(inputs[2]).toBeDefined();
+        expect(sortedClassList(inputs[0].nativeElement)).not.toContain('ng-submitted');
+        expect(sortedClassList(arrEl)).not.toContain('ng-submitted');
+        expect(sortedClassList(formEl)).toEqual(['ng-pristine', 'ng-untouched', 'ng-valid']);
```
I'd propose focusing the test on `ng-submitted` classes, so we can just check that we don't have in on the form elements:
```suggestion
        expect(sortedClassList(formEl)).not.toContain('ng-submitted');
```
Same applies to other tests below: we can just check for 'the ng-submitted' class presence.

dylhunn replied:
I have done this in most spots, except where the other classes were part of a pre-existing check.

AndrewKushnir commented on packages/forms/test/reactive_integration_spec.ts:
```diff
@@ -1123,6 +1138,126 @@ const ValueAccessorB = createControlValueAccessor('[cva-b]');
+        fixture.componentInstance.form = formGroup;
+        fixture.detectChanges();
+
+        const [loginInput, passwordInput] = fixture.debugElement.queryAll(By.css('input'));
```
nit: we should probably just get a hold of `nativeElement`s here (so we don't call it below):
```suggestion
        const [loginInput, passwordInput] = fixture.debugElement.queryAll(By.css('input')).map(el => el.nativeElement);
```

AndrewKushnir commented on packages/forms/test/reactive_integration_spec.ts:
```diff
@@ -4736,6 +4871,21 @@ class FormArrayComp {
+})
+class NestedFormArrayNameComp {
+  form!: FormGroup;
+  arr!: FormArray;
```
This field is unused (so we can remove it and avoid assigning it in tests), the `formArrayName="arr"` would be interpreted as a string input `arr` to the `formArrayName` directive.

AndrewKushnir reviewed (approved):
The changes look good, thanks @dylhunn 👍 

Just left a few comments related to tests.

atscott commented on packages/forms/src/directives/ng_control_status.ts:
```diff
@@ -80,10 +87,17 @@ export class NgControlStatus extends AbstractControlStatus {
 })
 export class NgControlStatusGroup extends AbstractControlStatus {
-  constructor(@Optional() @Self() cd: ControlContainer) {
+  constructor(@Optional() @Self() private cd: ControlContainer) {
```
I would suggest renaming this to `controlContainer` rather than `cd` now that it's used within the class and not just a one-off in the next line. `cd` often refers to "change detection" so the variable name is pretty confusing here.

AndrewKushnir replied:
+1, great idea. We should also rename it in the `NgControlStatus` directive above for consistency.

atscott commented on packages/forms/src/directives/ng_control_status.ts:
```diff
@@ -80,10 +87,17 @@ export class NgControlStatus extends AbstractControlStatus {
+
+  is(status: AnyGroupStatus): boolean {
+    if (status === 'submitted') {
+      return !!(this.cd as unknown as {submitted: boolean})?.submitted;
```
It's really hard to tell where the submitted property comes from. Is it possible to make this `as NgForm|null`? Or at least add a comment to describe where it comes from since it's not on the `ControlContainer` interface.

Either way, the `as` cast should include `null` since the control container is optional.

AndrewKushnir replied:
Yeah, I think comment is a good idea (and also adding `|null`). Since directive selector matches other types (that don't have `submitted` property) such as `FormArrayName` and `FormGroupName` directives, having more precise type might be tricky here (without introducing `instanceof` checks with non-tree-shakable references to these types).

atscott commented on goldens/public-api/forms/forms.d.ts:
```diff
@@ -377,6 +377,7 @@ export declare class NgControlStatus extends ɵangular_packages_forms_forms_i {
 
 export declare class NgControlStatusGroup extends ɵangular_packages_forms_forms_i {
     constructor(cd: ControlContainer);
+    is(status: AnyGroupStatus): boolean;
```
Just checking - this function name was more or less already public API because of the `extends AbstractControlStatus`, right? I think something like `hasStatus` would be better but there's not a ton that can be done if it's already done.

I'm a tad unsure about this since `AbstractControlStatus` isn't marked as public API and so it and its public methods don't show up in the [docs](https://angular.io/api/forms/NgControlStatusGroup). But since the method is public, I believe it would already be available (and public) even though it's not documented, right?

// cc @AndrewKushnir

AndrewKushnir replied:
Great point. Ideally it'd be great to avoid exposing that method altogether (including the method from the base class). So it's a tricky situation where the `is` method is technically a part of the public surface, but it's not in the docs (since base class is not marked as public API).

I've introduced the `is` method relatively recently (while doing  refactoring in https://github.com/angular/angular/commit/72c0188e05255a45e3e4662904cbbeef47013738), so I should've make it `protected` at that time :( Given that and the current state, I believe it might be safe to make it `protected` now, what do you think?

dylhunn replied:
I have marked this `protected`, which should be pretty low-risk, and I'll rerun the presubmits for any obvious issues.

@alxhub would you sanity-check that I'm not missing anything here?

dylhunn replied:
On second thought, I talked to @AndrewKushnir and we decided to completely remove the public API changes -- we can mark this protected and do some renaming/refactoring in the next major.

atscott commented on aio/content/guide/form-validation.md:
```diff
@@ -195,6 +195,7 @@ The following classes are currently supported.
 * `.ng-dirty`
 * `.ng-untouched`
 * `.ng-touched`
+* `.ng-submitted` (form groups and arrays only)
```
Based on what I'm seeing (searching for `.submitted =` in the forms package and based on the tests), arrays don't get the `ng-submitted` class.

atscott commented on packages/forms/src/directives/ng_control_status.ts:
```diff
@@ -13,6 +13,7 @@ import {ControlContainer} from './control_container';
 import {NgControl} from './ng_control';
 
 type AnyControlStatus = 'untouched'|'touched'|'pristine'|'dirty'|'valid'|'invalid'|'pending';
+type AnyGroupStatus = AnyControlStatus|'submitted';
```
Similar to my other comment - isn't the `submitted` only for forms, not just any group?

atscott commented on packages/forms/test/template_integration_spec.ts:
```diff
@@ -9,6 +9,7 @@
 import {ɵgetDOM as getDOM} from '@angular/common';
 import {Component, Directive, forwardRef, Input, Type, ViewChild} from '@angular/core';
 import {ComponentFixture, fakeAsync, TestBed, tick, waitForAsync} from '@angular/core/testing';
+import {expect} from '@angular/core/testing/src/testing_internal';
```
was this import intended?

 I _expect_ not :) :)  Someone's got to do it while @jessicajaniuk is away.

dylhunn commented on packages/forms/test/reactive_integration_spec.ts:
```diff
@@ -1123,6 +1138,126 @@ const ValueAccessorB = createControlValueAccessor('[cva-b]');
+        const form = new FormGroup({cities: cityArray});
+        fixture.componentInstance.form = form;
+        fixture.componentInstance.cityArray = cityArray;
+        cityArray.push(new FormControl('LA'));
```
@AndrewKushnir which line are you referring to?

AndrewKushnir replied:
I'm curious about this line:
```
cityArray.push(new FormControl('LA'));
```
Why do we add more elements into that FormArray?

dylhunn replied:
I see, it's not needed!

AndrewKushnir commented on packages/forms/src/directives/ng_control_status.ts:
```diff
@@ -80,10 +87,19 @@ export class NgControlStatus extends AbstractControlStatus {
+
+  is(status: AnyFormStatus): boolean {
+    if (status === 'submitted') {
+      // We check for the `submitted` field from FormArray FormGroup directives, but
```
```suggestion
      // We check for the `submitted` field from `NgForm` and `FormGroupDirective` classes, but
```

AndrewKushnir commented on packages/forms/test/reactive_integration_spec.ts:
```diff
@@ -1123,6 +1138,126 @@ const ValueAccessorB = createControlValueAccessor('[cva-b]');
+        cityArray.push(new FormControl('LA'));
+        fixture.detectChanges();
+
+        const inputs = fixture.debugElement.queryAll(By.css('input'));
```
nit: I was thinking if we can have something like this here as well?
```
        const [loginInput, passwordInput] =
            fixture.debugElement.queryAll(By.css('input')).map(el => el.nativeElement);
```
So we access the `nativeElement` just once here and the rest of the test should be a bit more easier to read.

AndrewKushnir commented on packages/forms/test/reactive_integration_spec.ts:
```diff
@@ -1123,6 +1138,126 @@ const ValueAccessorB = createControlValueAccessor('[cva-b]');
+
+        expect(sortedClassList(loginInput)).not.toContain('ng-submitted');
+        expect(sortedClassList(groupEl)).not.toContain('ng-submitted');
+        expect(sortedClassList(formEl)).toEqual(['ng-dirty', 'ng-invalid', 'ng-touched']);
```
Can we just check that there is no `ng-submitted` class here?
```
expect(sortedClassList(formEl)).not.toContain('ng-submitted');
```

atscott reviewed (approved):
👍  LGTM

dylhunn commented:
@atscott Andrew, I still need public API approval, would you mind approving with `reviewed-for: public-api`?

dylhunn commented:
After some design discussion and fiddling with the tests, Andrew and I decided that the best option for now is to simply extend the `is` method on `AbstractControlStatus`. Overriding it in `NgControlStatusGroup` is not currently desirable because that would add a public method to the interface, and expose it in the `.d.ts` file. It cannot be marked `private` or `protected` because that breaks ViewEngine tests. We should move it into the proper place after ViewEngine is removed.

This also undoes any real public API changes -- there is just one tiny no-op change in the `.d.ts`.

atscott reviewed (approved):
Reviewed-for: public-api

AndrewKushnir commented on packages/forms/src/directives/ng_control_status.ts:
```diff
@@ -22,6 +23,13 @@ export class AbstractControlStatus {
   }
 
   is(status: AnyControlStatus): boolean {
+    // This can be moved into NgControlStatusGroup after ViewEngine is removed; at present, it is
+    // not possible to make it protected, so adding it there would extend the public API.
```
```suggestion
    // Currently with ViewEngine (in AOT mode) it's not possible to use private methods in host bindings.
    // TODO: once ViewEngine is removed, this function should be refactored:
    //  - make the `is` method `protected`, so it's not accessible publicly
    //  - move the `submitted` status logic to the `NgControlStatusGroup` class
    //    and make it `private` or `protected` too.
```

AndrewKushnir reviewed (approved):
LGTM (just a minor comment) 👍

AndrewKushnir reviewed (approved):
Reviewed-for: public-api

jelbourn commented on aio/content/guide/form-validation.md:
```diff
@@ -195,6 +195,7 @@ The following classes are currently supported.
 * `.ng-dirty`
 * `.ng-untouched`
 * `.ng-touched`
+* `.ng-submitted` (forms only)
```
Why mark `.ng-submitted` as "forms only" here?

dylhunn replied:
Because the class will not be applied to individual controls, just the enclosing form element. I will clarify this a bit.

jelbourn commented on aio/content/guide/forms.md:
```diff
@@ -314,6 +314,8 @@ Angular sets special CSS classes on the control element to reflect the state, as
 
 </table>
 
+There is also an `ng-submitted` class which applies to the containing form when it is submitted. This class is *not* applied to inner controls.
```
"There is" is a bit passive voice here, it might be good to make it more active, something like

"Additionally, Angular applies the `ng-submitted` CSS class to `<form>` elements upon submission. This class does _not_ apply to inner controls.

jelbourn commented on packages/forms/test/reactive_integration_spec.ts:
```diff
@@ -1123,6 +1138,126 @@ const ValueAccessorB = createControlValueAccessor('[cva-b]');
+        expect(sortedClassList(formEl)).not.toContain('ng-submitted');
+      });
+
+      it('should work with nested formGroupName', () => {
```
Could this test name be more specific than "should work"? It would be good to be more explicit about what "work" means in this context

https://github.com/angular/angular/blob/master/docs/CODING_STANDARDS.md#test-names

AndrewKushnir replied:
Yeah, good point, there is a descriptive test name above (added in this PR) and we could use similar pattern for other tests:
https://github.com/angular/angular/pull/42132/files#diff-9d49b87bd9f813d36f5da177bfee5465c78e2e266129c989bd2035307f364e9dR1154

dylhunn replied:
Done, both for this and the above test.

jelbourn reviewed (approved):
LGTM for the public API change, just a few docs/naming nits

Reviewed-for: public-api

dylhunn commented:
Merging is currently blocked on a g3 presubmit issue, as described internally in b/189261737.

dylhunn commented:
Some updates:

- Real google3 breakage is fixed (see b/189261737 and cl/377141465)
- Some targets still fail to build in the presubmit; however these breakages are present in a clean client and can be ignored

jessicajaniuk commented:
This PR caused breakages in google3 and was rolled back. We'll need to run a TGP with these changes to verify there are no issues.

dylhunn committed:
feat(forms): undo revert and add ng-submitted class to forms that have been submitted. (#42132)

As previously discussed in pull/31070 and issues/30486, this would be useful because it is often desirable to apply styles to fields that are both `ng-invalid` and `ng-pristine` after the first attempt at form submission, but Angular does not provide any simple way to do this (although evidently Angularjs did). This will now be possible with a descendant selector such as `.ng-submitted .ng-invalid`.

In this implementation, the directive that sets control status classes on forms and formGroups has its set of statuses widened to include `ng-submitted`. Then, in the event that `is('submitted')` is invoked, the `submitted` property of the control container is returned iff it exists. This is preferred over checking whether the container is a `Form` or `FormGroup` directly to avoid reflecting on those classes.

Closes #30486.

PR Close #42132.

This reverts commit 00b1444d122a64a5b6772d694c838ce8b9ca7b7d, undoing the rollback of this change.

dylhunn commented:
The issue seems to have been that ViewEngine chokes on the spread operator as used in the host binding. I'll investigate this more separately, but it should be fixed here for now.

dylhunn commented:
After re-running flakes, TGP looks [good](https://test.corp.google.com/ui#id=OCL:377954179:BASE:378075315:1623128945944:2aeb2e0a) (googler link).

AndrewKushnir commented on packages/forms/src/directives/ng_control_status.ts:
```diff
@@ -36,6 +48,17 @@ export const ngControlStatusHost = {
   '[class.ng-pending]': 'is("pending")',
 };
 
+export const ngGroupStatusHost = {
```
@dylhunn could you please create a followup PR to add some information on why the content of the `ngGroupStatusHost` object partially duplicates the `ngControlStatusHost` one (related to spread operator support in ViewEngine and the need to refactor the code once ViewEngine is removed)?

dylhunn replied:
Will do.
